src/GraphQL/Validation/Complexity/ComplexityValidationRule.cs

```diff
@@ -0,0 +1,41 @@
+namespace GraphQL.Validation.Complexity;
+
+/// <summary>
+/// Analyzes a document to determine if its complexity exceeds a threshold.
+/// </summary>
+public class ComplexityValidationRule : IValidationRule
+{
+    private readonly ComplexityConfiguration _complexityConfiguration;
+    private readonly IComplexityAnalyzer _complexityAnalyzer;
+
+    /// <summary>
+    /// Initializes an instance with the specified complexity configuration.
+    /// </summary>
+    public ComplexityValidationRule(ComplexityConfiguration complexityConfiguration)
+        : this(complexityConfiguration, new ComplexityAnalyzer())
+    {
+    }
+
+    /// <summary>
+    /// Initializes an instance with the specified complexity configuration and complexity analyzer.
+    /// </summary>
+    public ComplexityValidationRule(ComplexityConfiguration complexityConfiguration, IComplexityAnalyzer complexityAnalyzer)
+    {
+        _complexityConfiguration = complexityConfiguration;
+        _complexityAnalyzer = complexityAnalyzer;
+    }
+
+    /// <inheritdoc/>
+    public ValueTask<INodeVisitor?> ValidateAsync(ValidationContext context)
+    {
+        try
+        {
+            _complexityAnalyzer.Validate(context.Document, _complexityConfiguration, context.Schema);
```

Comment:
I suggest to change Validate signature to return `ComplexityError(s)`.

Reply:
#3175